### PR TITLE
Rework composer autocomplete to be smarter and not trap tab

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/element-web/issues/4872
Fixes https://github.com/vector-im/element-web/issues/11071
Fixes https://github.com/vector-im/element-web/issues/17171
Fixes https://github.com/vector-im/element-web/issues/15646

Notes: Autocomplete has been updated to match modern accessibility standards. Navigate via up/down arrows rather than Tab. Enter or Tab to confirm a suggestion. This should be familiar to Slack & Discord users. You can now use Tab to navigate around the application and do more without touching your mouse. No more accidentally sending half of people's names because the completion didn't fire on Enter!

This PR brings our autocomplete in line with ARIA standards and matches the behaviour of Slack & Discord so should be easier for new users to comprehend.

|                 | Before               | After                |
|-----------------|----------------------|----------------------|
| Tab             | Navigate Suggestions | Accepts Suggestion   |
| Vertical Arrows | Navigate Suggestions | Navigate Suggestions |
| Enter           | Sends Message        | Accepts Suggestion   |

This means that the Tab key no longer keeps you trapped in the composer

Preview: https://6114f8642fe2a3998a4ce3cd--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Rework composer autocomplete to be smarter and not trap tab ([\#5659](https://github.com/matrix-org/matrix-react-sdk/pull/5659)). Fixes vector-im/element-web#4872, vector-im/element-web#11071 vector-im/element-web#17171 and vector-im/element-web#15646.

### CODE REVIEW

**Comment by SimonBrandner** [comment]
Fixes [#4872](https://github.com/vector-im/element-web/issues/4872)

**Comment by pvagner** [comment]
Unfortunatelly this no longer applies. I was using it for a while and now need to either fix this or revert and get used to the default behaviour.

**Comment by t3chguy** [comment]
@pvagner what do you mean it no longer applies?

**Comment by t3chguy** [comment]
Ah yes, github doesn't show merge conflicts for draft PRs annoyingly, I've resolved that for you now @pvagner 

The merge conflict was quite complex, I gave it a cursory test but let me know if it behaves whatsoever differently, it shouldn't.

Thanks!

File: src/components/views/rooms/BasicMessageComposer.tsx [hunk 1]
```diff
@@ -567,7 +577,7 @@ export default class BasicMessageEditor extends React.Component<IProps, IState>
 
     private onAutoCompleteSelectionChange = (completion: ICompletion, completionIndex: number) => {
         this.modifiedFlag = true;
-        this.props.model.autoComplete.onComponentSelectionChange(completion);
+        // this.props.model.autoComplete.onComponentSelectionChange(completion);
```
Comment (turt2live) [inline src/components/views/rooms/BasicMessageComposer.tsx:580]: but whyyyy....

(add comment or remove please)

**Review by turt2live — APPROVED** [review]
looks good overall

File: src/components/views/rooms/BasicMessageComposer.tsx (hunk 1 above)
Comment (t3chguy) [reply src/components/views/rooms/BasicMessageComposer.tsx:580]: Oooh this pointed out more code I can 🗑️

**Review by turt2live — APPROVED** [review]
`rm -rf tabcomplete`

**Review by nadonomy — APPROVED** [review]
LGTM.

@novocaine @dbkr In the release notes, I think we should highlight this change as existing users will need to battle existing muscle memory and it's bound to cause friction.

I think we'd want to say something along the lines of "Autocomplete suggestions have been updated to match modern accessibility standards. Navigate via arrows rather than tab. Enter to confirm a suggestion."

**Comment by novocaine** [comment]
> LGTM.
> 
> @novocaine @dbkr In the release notes, I think we should highlight this change as existing users will need to battle existing muscle memory and it's bound to cause friction.
> 
> I think we'd want to say something along the lines of "Autocomplete suggestions have been updated to match modern accessibility standards. Navigate via arrows rather than tab. Enter to confirm a suggestion."

note that thanks to the new changelog automation, @t3chguy was simply able to update the PR description with your copy, and dave and I didn't need to lift a finger

**Comment by MTRNord** [comment]
Can we get this as  a setting instead? This is a breaking change without warning and it gets annoying already. :/ Tab is the correct way but this breaks it so I now need to add unneeded spaces to be able to send messages as well as taking 2x to write a message. Also there are 8/10 times the wrong emoji being autocompleted. For example `:O` will get autocompleted to `:v:` which is no where near the correct one.

**Comment by turt2live** [comment]
Please open an issue to report problems. PR comments are not monitored for triage.

**Pull request merged**